api/algorithms/group_manager.py

```diff
@@ -0,0 +1,148 @@
+# DO NOT USE IN PROD, WORK IN PROGRESS
+
+from collections.abc import Iterable, Iterator
+from dataclasses import dataclass
+from datetime import datetime
+
+from django.utils import timezone
+
+from api.models import AgeRange, Group, Student, Teacher
+from api.models.auxil.constants import (
+    DEFAULT_LESSON_DURATION_MIN,
+    MAX_AGE_KIDS_GROUP,
+    MAX_AGE_TEEN_GROUP,
+)
+from api.models.auxil.status_setter import StatusSetter
+from api.models.choices.communication_language_mode import CommunicationLanguageMode
+from api.models.choices.log_event_type import (
+    GroupLogEventType,
+    StudentLogEventType,
+    TeacherLogEventType,
+)
+from api.models.choices.status import GroupStatus, TeacherStatus
+from api.models.language_and_level import LanguageAndLevel
+from api.processors.auxil.log_event_creator import GroupLogEventCreator
+
+
+class GroupManager:
+    @dataclass
+    class GroupSizeRestriction:
+        min_number: int
+        max_number: int
+
+    @dataclass
+    class GroupCandidate:
+        language_and_level: LanguageAndLevel
+        communication_language_mode: CommunicationLanguageMode
+        age_range: AgeRange
+        teacher: Teacher
+        students: list[Student]
+        monday: datetime | None = None
+        tuesday: datetime | None = None
+        wednesday: datetime | None = None
+        thursday: datetime | None = None
+        friday: datetime | None = None
+        saturday: datetime | None = None
+        sunday: datetime | None = None
+
+        def is_acceptable(self) -> bool:
+            # Checks group size is appropriate for age range
+            restrictions = GroupManager._get_allowed_group_size(self.age_range)
+            students_num = len(self.students)
+            if restrictions.min_number > students_num:
+                return False
+            if restrictions.max_number < students_num:
+                raise ValueError(
```

Comment:
Why not return False here? Or why not raise ... in case of `restrictions.min_number > students_num` ?

Reply:
We were having the same discussion [here](https://github.com/samanthasgroup/django-webapps/pull/209#discussion_r1285262948), so why don't you join us 😄

Reply:
Hiding this as a duplicate